# Trying New Features and features selection

In [1]:
import pandas as pd
train_df=pd.read_csv('train_dataset.csv')
test_df=pd.read_csv('test_dataset.csv')

In [2]:
train_df.shape,test_df.shape

((16500, 22), (3500, 22))

In [3]:
corr_list=train_df.astype(float).corr()['Golden Grains'].sort_values(axis=0,ascending=False)

In [4]:
corr_list.abs().sort_values(axis=0,ascending=False)

Golden Grains         1.000000
capital_dist          0.875949
knight_house_dist     0.875924
blessing              0.846229
royalMarket_dist      0.213089
location              0.183795
dock_dist             0.160422
Built_Date            0.150510
bedroom               0.149919
Priced_date           0.135443
guardingTower_dist    0.132951
river_dist            0.129185
dining                0.124203
farm_land             0.082550
Garden                0.063837
kingVisit             0.060831
curse                 0.048853
holy_tree             0.033636
bathroom              0.024816
renovation            0.007550
House ID              0.003668
builder               0.000226
Name: Golden Grains, dtype: float64

Checking correlation of features with target value

In [5]:
train_df['area']=train_df['dining']+train_df['bedroom']+train_df['kingVisit']+train_df['farm_land']

In [6]:
corr_list=train_df.astype(float).corr()['Golden Grains'].sort_values(axis=0,ascending=False)
corr_list.abs().sort_values(axis=0,ascending=False)

Golden Grains         1.000000
capital_dist          0.875949
knight_house_dist     0.875924
blessing              0.846229
royalMarket_dist      0.213089
area                  0.205306
location              0.183795
dock_dist             0.160422
Built_Date            0.150510
bedroom               0.149919
Priced_date           0.135443
guardingTower_dist    0.132951
river_dist            0.129185
dining                0.124203
farm_land             0.082550
Garden                0.063837
kingVisit             0.060831
curse                 0.048853
holy_tree             0.033636
bathroom              0.024816
renovation            0.007550
House ID              0.003668
builder               0.000226
Name: Golden Grains, dtype: float64

This new engineered feature gave 20.5% correlation, lets look how it affects dataset

In [7]:
train_df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 16500 entries, 0 to 16499
Data columns (total 23 columns):
House ID              16500 non-null int64
Built_Date            16500 non-null int64
Priced_date           16500 non-null int64
Garden                14515 non-null float64
dock_dist             14859 non-null float64
capital_dist          15669 non-null float64
royalMarket_dist      14146 non-null float64
guardingTower_dist    15956 non-null float64
river_dist            16500 non-null float64
renovation            14871 non-null float64
dining                14862 non-null float64
bedroom               15851 non-null float64
bathroom              15536 non-null float64
kingVisit             16500 non-null int64
curse                 15502 non-null float64
blessing              16500 non-null int64
farm_land             16500 non-null int64
location              14829 non-null float64
holy_tree             13234 non-null float64
knight_house_dist     15669 non-null float64
bui

This produces lots of NaN values which might be difficult to fill and also this feature is not so profitable so better to be dropped

In [8]:
train_df.drop('area',axis=1)

House ID  Built_Date  Priced_date  Garden   dock_dist  capital_dist  \
0      1848823488        1602         1609     0.0  138.600639      0.256239   
1      1848823492        1600         1612     1.0   16.254871     26.882222   
2      1848823497        1600         1608     0.0    6.555428    136.448973   
3      1848823502        1603         1611     0.0   79.724550     19.885265   
4      1848823505        1603         1610     0.0  101.541256      0.519507   
5      1848823507        1603         1612     0.0   48.601612      7.575922   
6      1848823509        1600         1607     0.0   32.624227      1.955970   
7      1848823511        1603         1610     0.0   84.170405     15.776747   
8      1848823514        1603         1611     0.0   10.998322    122.930753   
9      1848823517        1601         1607     0.0   60.265361      2.809867   
10     1848823519        1601         1609     0.0   11.938817      2.738666   
11     1848823524        1601         1607     0.0   13.986232      0.212412   
12     1848823529        1603         1611     NaN   38.977167      5.128739   
13     1848823531        1600         1611     0.0  135.364314     85.723029   
14     1848823533        1601         1612     0.0   23.614598     24.762207   
15     1848823539        1602         1611     0.0         NaN     23.338276   
16     1848823545        1601         1611     0.0   35.488288     68.371189   
17     1848823548        1603         1608     0.0   57.719253     67.341891   
18     1848823550        1602         1612     0.0   43.156050     66.593962   
19     1848823553        1603         1607     0.0   81.589930     52.209955   
20     1848823559        1606         1612     0.0         NaN     20.748392   
21     1848823561        1603         1610     0.0   52.819705     80.970959   
22     1848823566        1606         1609     0.0   60.198424     65.509891   
23     1848823568        1600         1609     0.0   31.038516      0.531688   
24     1848823571        1605         1608     0.0   85.472652     70.864401   
25     1848823575        1603         1610     NaN   18.692850      1.034696   
26     1848823578        1605         1612     0.0   11.585549     54.728789   
27     1848823584        1600         1610     0.0  145.034612     25.106037   
28     1848823590        1600         1607     0.0    1.072398     25.097674   
29     1848823594        1606         1607     0.0   27.395888     16.519207   
...           ...         ...          ...     ...         ...           ...   
16470  1848889275        1602         1610     0.0         NaN     36.319084   
16471  1848889281        1600         1607     0.0    5.742986     15.209968   
16472  1848889287        1603         1608     0.0   20.913353     67.385719   
16473  1848889291        1606         1610     0.0   69.716999     19.279870   
16474  1848889296        1601         1607     0.0   43.759441     69.772198   
16475  1848889300        1600         1608     NaN  139.070334     23.953692   
16476  1848889304        1600         1607     NaN   25.336431     36.051541   
16477  1848889309        1600         1608     0.0   24.982941     35.056489   
16478  1848889315        1606         1610     0.0   79.570596     86.823738   
16479  1848889317        1606         1608     0.0   75.595337     68.192290   
16480  1848889319        1603         1608     0.0   64.299897     19.104478   
16481  1848889323        1604         1611     0.0   36.629314    108.163721   
16482  1848889326        1605         1612     NaN   32.751357     57.235709   
16483  1848889328        1600         1609     0.0         NaN     45.507866   
16484  1848889334        1605         1608     0.0    1.143295     47.292367   
16485  1848889337        1606         1612     1.0         NaN     13.572429   
16486  1848889342        1603         1610     0.0   25.339922     59.696311   
16487  1848889345        1603         1607     0.0    0.952657     14.783799   
16488  18

In [9]:
train_df.interpolate().count()

House ID              16500
Built_Date            16500
Priced_date           16500
Garden                16500
dock_dist             16500
capital_dist          16500
royalMarket_dist      16500
guardingTower_dist    16500
river_dist            16500
renovation            16500
dining                16500
bedroom               16500
bathroom              16498
kingVisit             16500
curse                 16500
blessing              16500
farm_land             16500
location              16500
holy_tree             16500
knight_house_dist     16500
builder               16500
Golden Grains         16500
area                  16500
dtype: int64

In [10]:
train_df.interpolate().plot()

through interpolate we will fill NaN values as they are far better than fillna() of pandas in some ways

In [11]:
train_df.interpolate(method='akima',inplace=True)

‘krogh’, ‘piecewise_polynomial’, ‘spline’, ‘pchip’ and ‘akima’ are all wrappers around the scipy interpolation methods of similar names. These use the actual numerical values of the index. 

In [12]:
train_df['bathroom'].fillna(2,inplace=True),train_df['bathroom'].fillna(2,inplace=True)

(None, None)

In [13]:
test_df.interpolate(method='akima',inplace=True)

In [14]:
test_df['location'].fillna(1,inplace=True),test_df['renovation'].fillna(1,inplace=True)

(None, None)

These are the values which remain unfilled by interpolation

In [15]:
train_df.drop(['Garden','House ID','renovation'],axis=1,inplace=True)
test_df.drop(['Garden','House ID','renovation'],axis=1,inplace=True)

These features have least correlation with target so better to be dropped

# Now our data is ready to be used